### DISCUSSION TITLE
Using a HashableArrayWrapper inside JIT'ted function fails

### DISCUSSION THREAD

**Discussion by srijaniiserprinceton** [jax-ml/jax#8439] [Q&A] [answered]
I am still in the process of learning how to think in JAX. I have been using a `HashableArrayWrapper` to pass arrays as static arguments to functions along the lines of [this answer](https://github.com/google/jax/issues/4572#issuecomment-709677518). This was working for me quite well. But for some reason that I don't quite understand, it has been failing in this [script](https://colab.research.google.com/drive/1Qhnv1PoOwh3QpL8oGWZ5eoPXptmt9YNA?usp=sharing#scrollTo=PoyKj4ZOxEdv). This collab script should be accessible by anyone. 

I decorate the function `calc_rad_from_coor` with `gnool_jit` (that lets me pass hashable arrays as static arguments) which is called from within the function `model`.
```
@partial(gnool_jit, static_array_argnums=(0,1))
def calc_rad_from_coor(x_m, y_m):
    return np.sqrt(x_m**2 + y_m**2)

def model(r_model):
    npts = len(theta_d)
    
    x_model = r_model * np.cos(theta_d)
    y_model = r_model * np.sin(theta_d)
    
    # adding gjit function here                                                                
    r_model_gjit = calc_rad_from_coor(x_model, y_model)
    
    mu = r_model - r_d
    return np.sum(-0.5*(mu/sigma_d)**2)
```
Note that `model` is called from within another `jax.jit`ted function `rw_metropolis_kernel`. 

I have, in the past used `gnool_jit`ted functions within `jax.jit`ted functions but it had worked. Not sure what is happening here. :(

**Comment by jakevdp** [comment, marked as answer]
I don't entirely understand the motivation for your `gnool_jit`, but within a jit-compiled function, the result of something like `x_model = r_model * np.cos(theta_d)` is not an array, it is a tracer. When you pass this tracer to a function that expects it to be a static array, the result will probably not work as expected.

**Reply by srijaniiserprinceton** [reply]
So, the exact error I get is the following:
```
UnfilteredStackTrace: ValueError: Non-hashable static arguments are not supported. An error occured while trying to hash an object of type <class '__main__.HashableArrayWrapper'>, <__main__.HashableArrayWrapper object at 0x7f6be0041150>. The error was:
ConcretizationTypeError: Abstract tracer value encountered where concrete value is expected: Traced<ShapedArray(float64[])>with<BatchTrace(level=1/1)>
  with val = Traced<ShapedArray(float64[4])>with<DynamicJaxprTrace(level=0/1)>
       batch_dim = 0
The problem arose with the `int` function. If trying to convert the data type of a value, try using `x.astype(int)` or `jnp.array(x, int)` instead.
```
As far as I understand, this happens at the `some_hash_function` where we are trying to cast the sum of the array to an int (since hash values need to be int)
```
def some_hash_function(x):
    return np.sum(x.astype('int'))

class HashableArrayWrapper:
    def __init__(self, val):
        self.val = val
    def __hash__(self):
        return some_hash_function(self.val)
    def __eq__(self, other):
        return (isinstance(other, HashableArrayWrapper) and
                np.all(np.equal(self.val, other.val)))

def gnool_jit(fun, static_array_argnums=()):
    @partial(jax.jit, static_argnums=static_array_argnums)
    def callee(*args):
        args = list(args)
        for i in static_array_argnums:
            if isinstance(args[i], tuple):
                args[i] = args[i].__class__(*[a.val for a in args[i]])
            else:
                args[i] = args[i].val
        return fun(*args)

    def caller(*args):
        args = list(args)
        for i in static_array_argnums:
            if isinstance(args[i], tuple):
                all_as = [HashableArrayWrapper(a) for a in args[i]]
                args[i] = args[i].__class__(*all_as)
            else:
                args[i] = HashableArrayWrapper(args[i])
        return callee(*args)

    return caller
```
    
Is there anyway to get around this? All I want to do is be able to send arrays as static arguments. And the decorator `gnool_jit` which allows this fails inside a jit'ted function. The code can be found and executed in [this Collab script](https://colab.research.google.com/drive/1Qhnv1PoOwh3QpL8oGWZ5eoPXptmt9YNA?usp=sharing).

**Reply by jakevdp** [reply]
I can't think of any way to do that.

The good news is, I suspect this is an [XY problem](https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem): that is, I doubt your high-level goal is to dispatch functions based on hashable wrappers of traced arguments. But your higher-level goal is probably possible. Setting aside hashable array wrappers in jit, can you describe your overall goal here?

**Reply by srijaniiserprinceton** [reply]
I definitely think I am falling into an XY-problem here! So, the overall goal is to use the Bayesian sampling provided by [Numpyro](http://num.pyro.ai/en/stable/getting_started.html). To do that we need to define a `model` function which `numpyro` essentially `jit`s since its going to run `model` many many times. 

Now, our `model` function involves two important steps:
```
def f1(<some args>):
    <calculates the indices to tile a big array in f2>
    <returns these indices>
    # as an example
    example_arr = [[10,20], [5, 19]]
    return example_arr

def f2(<args containing array of indices from f1>):
    <creates small matrices of shapes derived from the array calculated in f1>
    <tiles a big matrix using these small matrices by slicing the big matrix>
    <returns the big matrix>
    # as an example
    big_matrix[10:20, 5:19] = small_matrix  # of shape 10x14
    return big_matrix

def model(args_f1):
    a = f1(args_f1)     # a is the array of matrix indices that we will use to slice a big array inside f2
    b = f2(a)        # this is where the argument needs to be static so the 
                           # we can use it to create matrices and tile them in a big matrix inside f2
    return b
```
All we want to do is pass these arrays containing indices to f2 which are used to slice matrices and fill them. But doing so naively doesn't work since these arrays should be static since its elements are used to define shapes of matrices inside `f2`. So, we attempt to make the arrays static and therein comes the need to pass static arrays. 

So, in short, we want to make our `model` function `jit`able and inside that we want to `jit` the functions f1 and f2 which will take in arrays of indices. This array of indices will be used to create smaller matrices and tile a larger matrix using the smaller matrices.

**Reply by jakevdp** [reply]
OK, that makes sense. In that case, I don't think the HashableArrayWrapperr is a great approach - for example, it hashes based on the sum of the array. I'd bet you could come up with two arrays that have the same sum and result in different tilings, so that's not a great hash function!

In your case, since the value of every element of the array matters and must be static, have you considered converting it to nested tuples of integers? Then it could be hashed based on the value without any special wrapper, which is probably what you want here. And if you avoid using `jnp` functions to define or manipulate your array of shapes, it will remain static and you can pass the tuple of tuples to a standard jitted function.

Does that make sense?

**Reply by srijaniiserprinceton** [reply]
Oh wow! Never even thought of using nested tuples for that. They indeed have a `__hash__` function. I'll get back to you on this within a day after replacing the arrays with nested tuples and report if this worked. Fingers crossed!

**Reply by srijaniiserprinceton** [reply]
OK. I started converting the arrays to tuples and nested tuples, although there seems to be a minor obstacle in that. I have summarized the problem using a simple script
```
from jax import jit
import numpy as np

def f(a_tuple):
    # converting to array to perform usual ndarray ops
    a_arr = np.asarray(a_tuple)
    a_arr += 1
    # converting back to nested tuple
    a_tuple_ret = tuple(map(tuple,a_arr))
    # checking what it looks like
    print(a_tuple_ret)
    return a_tuple_ret

f_ = jit(f, static_argnums = (0,))

# the tuple
a_tuple = ((1,2),(3,4))

a_tuple_final = f_(a_tuple)

# every element is now a DeviceArray which prevents obtaining a hash value
print(a_tuple_final)
```
Running this code gives me 
```
((2, 3), (4, 5))
((DeviceArray(2, dtype=int32), DeviceArray(3, dtype=int32)), (DeviceArray(4, dtype=int32), DeviceArray(5, dtype=int32)))
```
When I try to obtain the hash value of the output `a_tuple_final` which is supposed to be the static input to another `jit`ted function, I get the following error as expected from a `DeviceArray`
```
a_tuple_final.__hash__()
TypeError: unhashable type: 'DeviceArray'
```
Is there a workaround for this? I have a feeling we need to use `jax.tree_util.tree_map` for this but not sure exactly how to.

**Reply by srijaniiserprinceton** [reply]
Update: One solution I figured out was this
```
import jax.tree_util as tu
a_tuple_final = tu.tree_map(lambda x: int(x), a_tuple_final)
```
At this point this has a hash value
```
print(a_tuple_final)
print(a_tuple_final.__hash__())
```
gives me 
```
((2, 3), (4, 5))
-3294401797587171123
```
I guess if its a tuple of `float`, we just replace `int` by `float` in the `tree_map`.

If you think this is not the most optimal way, please let me know.

**Reply by srijaniiserprinceton** [reply]
Using namedtuples instead of the previous method of defining hashable arrays works great! So, that essentially addresses the question. This *almost* solves the problem. But I do face a weird issue which I can't explain. I find that static arguments, even when acted upon only by `numpy` operations inside the `jit`ted `model` function become traced. The `model` function is as below and the [simplified script can be found here](https://colab.research.google.com/drive/17ImOoc_JrkPe1yxXaKLUFKBuXUTMsiwL?usp=sharing).
```
def model(GVARS):
    for i in range(1):
        print('\n Starting loop')
        print(GVARS)      # GVARS is still static here
        # calling the function with two static arguments
        a_static, b_static = func1_w_staticargs(GVARS.a, GVARS.b)
        # a_static and b_static become traced here
        # even thought we declare the argument to be static and use numpy ops only
        print('a_static, b_static: \n', a_static, b_static)
        
        # converting to arrays to tuple to pass as static arguments
        # to the next function call 
        a_static_tuple = tuple(map(tuple, a_static))
        b_static_tuple = tuple(b_static)
        GVARS_new_ = namedtuple('GVARS_new',
                               ['a','b'])
        GVARS_new = GVARS_new_(a_static_tuple, b_static_tuple)
        
        print('Printing as tuples: \n', GVARS_new.a, GVARS_new.b)
        print('\n')
        # print('Hash values :,\n', GVARS_new.a.__hash__(), GVARS_new.b.__hash__())

        # second function called with tuples passed as static arguments
        a_static, b_static = func2_w_staticargs(GVARS_new.a, GVARS_new.b)

        fac = 1.0
        eig_sample = np.asarray(GVARS_new.b)

    # eig_sample = numpyro.deterministic('eig', eig_mcmc_func(w1=w1, w3=w3, w5=w5))
    # return numpyro.sample('obs', dist.Normal(eig_sample, np.ones_like(b)), obs=np.asarray(b))
    return eig_sample
```
Note that I declare the argument in the `model` as static using `model_ = jax.jit(model, static_argnums=(0,))`.

Output:
```
Starting loop
GVARS(a=((1, 2), (3, 4)), b=(1.0, 2.0, 3.0, 4.0))
a_static, b_static: 
 [[2 3]
 [4 5]] [2. 3. 4. 5.]
Printing as tuples: 
 ((2, 3), (4, 5)) (2.0, 3.0, 4.0, 5.0)


[no-jit] model() ran successfully.

 Starting loop
GVARS(a=((1, 2), (3, 4)), b=(1.0, 2.0, 3.0, 4.0))
a_static, b_static: 
 Traced<ShapedArray(int64[2,2])>with<DynamicJaxprTrace(level=0/1)> Traced<ShapedArray(float64[4])>with<DynamicJaxprTrace(level=0/1)>
Printing as tuples: 
 ((Traced<ShapedArray(int64[])>with<DynamicJaxprTrace(level=0/1)>, Traced<ShapedArray(int64[])>with<DynamicJaxprTrace(level=0/1)>), (Traced<ShapedArray(int64[])>with<DynamicJaxprTrace(level=0/1)>, Traced<ShapedArray(int64[])>with<DynamicJaxprTrace(level=0/1)>)) (Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=0/1)>, Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=0/1)>, Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=0/1)>, Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=0/1)>)

UnfilteredStackTrace: ValueError: Non-hashable static arguments are not supported. An error occured while trying to hash an object of type <class 'tuple'>, ((Traced<ShapedArray(int64[])>with<DynamicJaxprTrace(level=0/1)>, Traced<ShapedArray(int64[])>with<DynamicJaxprTrace(level=0/1)>), (Traced<ShapedArray(int64[])>with<DynamicJaxprTrace(level=0/1)>, Traced<ShapedArray(int64[])>with<DynamicJaxprTrace(level=0/1)>)). The error was:
TypeError: unhashable type: 'DynamicJaxprTracer'
```

**Reply by srijaniiserprinceton** [reply]
Update: I removed all the `jit`s for the internal functions inside `model` and converted `jnp` ops on static objects to `np` ops and it works fine. I guess, the big takeaway for me personally is that I don't necessarily need to `jax.jit` (and declare static arguments) every internal function inside a larger function that is `jit`ted. Careful application of `numpy` instead of `jax.numpy` on all the static objects should do the job.

**Reply by jakevdp** [reply]
That makes sense.

Stepping back, though, it feels like you're spending a lot of time working against the JAX model rather than working with it. Is there a reason your arrays need to be static? I generally think of that as an anti-pattern in JAX. This is one reason there's not much built-in support for manipulation of static arrays: it's a bit counter to JAX's entire computation model.

**Reply by srijaniiserprinceton** [reply]
The arrays that I have been referring to (passing as static arguments to subsequent `jit`ted functions), are arrays containing dimensions of matrices (now changed to nestedtuples of ints). Meaning, the int values in these arrays are used in various places in my code to create shapes of arrays. So, every time this array (now nestedtuples) change, we need to recompile. But I am all ears for a different approach to doing this! But we do gain out of using JAX since we are doing a Bayesian simulation where N-1 iterations don't need re-compilations. But from what we observe, the time to compile (in the first iteration) is of the same order of magnitude as the time of executing the N-1 iterations. 

P.S: We have finally got our code to run in `Numpyro` (which just required us to ensure that the `model` function, which encapsulates all the other functions we are using, is `jit`able). However, our compile time when running through numpyro is quite long (of the order of hours). [The Numpyro dev asked us to talk to JAX devs about this and report the same](https://forum.pyro.ai/t/how-does-run-time-on-setting-num-chains-scale-with-number-of-cores/3536?u=srijanbharatidas). Will report more on this issue soon!

**Reply by jakevdp** [reply]
I see, thanks. It's hard to say without seeing more, but I think for passing around shapes, tuples of tuples probably makes more sense than arrays.

As for the long compile times, are you using a lot of Python control flow? Be aware that if you JIT-compile functions containing Python for loops, those loops will be flattened and compilation tends to scale with about the square of the number of flattened operations.

**Reply by srijaniiserprinceton** [reply]
@jakevdp yep, that's precisely the culprit I think. As I mentioned above, we removed all the `partial` decorators for the functions where we pass the static arguments but make sure that we only do usual python and `numpy` ops on the static arguments so that they are never converted to dynamic traced values. This causes us to retain a lot of `np` operation and usual pythonic `for` loops. 

The reason we were forced to do this is because somehow when calling a `jit`-decorated function within `model` (which is the outermost jitted function), the returned values are traced (which we don't expect since they depend on static arguments). 

It would be great you could this [Collab notebook](https://colab.research.google.com/drive/17ImOoc_JrkPe1yxXaKLUFKBuXUTMsiwL?usp=sharing) which contains a simplified example of the overall structure of the code that we want to work and let us know how to make the very last call to `model_jnp_(GVARS)` functional (which, as you will see fails since the arguments that are getting passed to a subsequent function is getting traced).

**Reply by jakevdp** [reply]
Thanks. One comment on a read-through of the notebook. In these lines:
```python
        # calling the function with two static arguments
        a_static, b_static = func1_w_staticargs_jnp(GVARS.a, GVARS.b)

        # when jitted, a_static and b_static become traced here
        # even thought we declare the argument to be static and use numpy ops only
        print('a_static, b_static: \n', a_static, b_static)
```
The function `func1_w_staticargs_jnp` creates DeviceArrays, and any device array created within a JIT context will be traced. From the comment, it seems like maybe you meant to call `func1_w_staticargs_np`? If that were the case, I think the outputs would still be static.

**Reply by srijaniiserprinceton** [reply]
What I did in the collab notebook was to demonstrate why we are using `numpy` ops in our code which is making the compilation time long. So, it was intentional to call `func1_w_staticargs_jnp`. We discussed in your previous reply how non-`jax` operations (loops etc) are rolled out and then square of the number of operations are performed. So, in the notebook I was trying to show you why we are forced to use the `numpy` ops (in `func1_w_staticargs_np`) and why using the `jax.numpy` counterpart (in `func1_w_staticargs_jnp`) fails. It will be possible to stop the rolling out of loops in our actual code, if a function like `func1_w_static_jnp` which is decorated with `static_argnums` returns static objects (in which case we can replace the `numpy` counterpart with the `jax.numpy` counterpart. 

So, in short:

- We currently use `numpy` ops on static arguments -> Makes compilation time long. (So what you wrote in your suggestion in the above reply is what we are doing now in the actual code.
- We are forced to do the above since the `jax.numpy` ops returns us traced objects which cannot be passed into a subsequent function that wants to use these values for creating shaped matrices.

**Reply by srijaniiserprinceton** [reply]
One of the options that I can think of to bypass this passing of static arguments is described below:

- I fix the shapes of the matrices that I need to make to the largest matrix that I would ever encounter. Let's call this large matrix `M_large`.
- I pass the nestedtuples (or arrays) which are traced (not static) to slice out and fill in smaller areas in the large matrix `M_large`. Is this possible by `dynamic_slice`? 

If it is possible to fill out smaller patches in `M_large` (the size of these small patches being governed by a traced value), then we can perhaps exclude static arguments altogether and make everything `numpy` instead of `jax.numpy`.

**Reply by jakevdp** [reply]
Oh, sorry, I got distracted by the comment, which says that `np` ops are being used, when it is `jnp` ops that are being used. Maybe it's just outdated.
